### PR TITLE
Repair interrupted interactive chat transcripts

### PR DESCRIPTION
**Pull request by tonyespinoza1**
## Why

Interactive chat persists transcript messages incrementally. If a process stops or a user cancels after an assistant tool call is recorded but before every tool result is durable, the next OpenAI Responses request replays a malformed call/result exchange and fails with HTTP 400. The original prefix-truncation repair also lost the tool-attempt record, later valid history, and provider compaction state.

## What Changed

- Use one non-destructive transcript normalizer at both durable restore and the next-turn boundary.
- Add an explicit tool result whose outcome is `unknown` for each missing result, without claiming the tool succeeded, failed, ran, or produced side effects.
- Preserve the assistant call, completed results from a partial parallel batch, later history, and provider continuation/compaction state.
- Delete the destructive truncation path.
- Fail closed locally before provider traffic for orphan or duplicate tool results and duplicate call IDs, leaving durable evidence unchanged.
- Document the interactive recovery boundary.

## Test plan

- Red-first regressions covered interrupted restart, cancel during tool execution, mid-history partial batches, provider compaction preservation, complete exchanges, malformed legacy histories, and normalization idempotence across a persisted restart.
- `deno task --no-lock test` in `packages/cf-harness`: 768 passed (444 steps).
- Pinned Deno 2.9.4 `deno task --no-lock check`: 43 package groups passed.
- Pinned Deno 2.9.4 `deno task --no-lock check-docs`: 566 code blocks passed.
- `deno lint` on changed TypeScript files.
- `deno fmt --check` on all changed files.
- Fresh LCOV confirms every service line added by this PR is exercised.
- `git diff --check`.

## Codex task

- Title: 🛠️ Fixing Labs PR 6325
- Thread ID: `01a03b44-52b8-7c63-8ff3-da8157912965`

<a href="https://cubic.dev/pr/commontoolsinc/labs/pull/6325?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by tonyespinoza1** [comment]
CI note: this head has three red lanes. `Check` and `Test (8/8)` reproduce the same `check-completion-slots` failure on the immediately preceding main run (missing candidates/allowlist entries for `piece survey --diff` and `piece retarget --group-size`); this PR cannot reach that CLI surface. `Runner Tests (1/8)` exhausted the existing bounded structural-window retry in `executor-events-down.test.ts`; main passed that lane and this PR does not reach runner. I am holding the merge, not absorbing either unrelated failure, and will refresh/re-run after main is green. Cubic found no issues; the local cf-harness package suite remains 762 passed.

**Comment by tonyespinoza1** [comment]
## Red team review

I cloned this branch and wrote two executable repros. The two high-severity findings below are **reproduced, not inferred** — repro sources are inlined at the bottom and drop straight into `packages/cf-harness/test/`.

### What holds up

Ran `truncateIncompleteToolExchange` directly over an edge-case matrix: partial batches truncate correctly, complete batches pass through untouched, out-of-order tool results are accepted, duplicate results are trimmed, and the `view_image` followup `user` message (`prompt-loop.ts:3154`) is pushed *after* the batch closes so it doesn't trip a false positive. The atomicity claim checks out — `saveSessionAndAppendEvent` wraps in `BEGIN IMMEDIATE` (`sqlite-session-store.ts:217`). Recovery is idempotent across repeated restarts.

---

## H1 — `cancel_turn` produces the identical corruption, and the new guard permanently excludes it

`cancelTurn` populates `canceledTurnIds`, after which `onTranscriptEvent` early-returns (`interactive-chat-service.ts:1152`) and the catch block early-returns (`:1176`). `record.transcript` therefore freezes at whatever was last persisted — and cancelling *during tool execution* is the most common moment to cancel, so it freezes on a stranded assistant tool-call that the `tool_started` emit already wrote durably.

The turn then settles as `canceled`, which `isTerminalTurnStatus` treats as terminal. So `hasInterruptedTurn` (`:714`) is **false** and the repair never runs — not on that restart, not on any later one. The session stays `reusable: true` and every subsequent turn 400s. `toResponsesInput` has no request-time call/result matching pass, so nothing downstream rescues it.

Output from the repro, run on this branch:

```
[3] session reusable: true
[4] durable transcript AFTER PR 6325 recovery: [ "user(Delete the temp files)", "assistant[call-canceled]" ]
[5] transcript sent to the model on the NEXT turn: [ user(...), assistant[call-canceled], user(Continue) ]
[6] UNMATCHED tool calls in next model input: [ "call-canceled" ]
```

The existing cancel test (`interactive-chat-service.test.ts:705`) never emits a tool call before cancelling, so this path has no coverage today.

---

## H2 — The repair truncates at the *first* defect, destroying all later history

The doc comment is literally accurate ("longest prefix which does not strand a batch"), but the strategy is truncate-to-defect rather than excise-defect. For a trailing defect those coincide; for a mid-transcript defect the blast radius is the entire session. Seeding a transcript with one early stranded call followed by three complete, valid turns:

```
[before] 13 messages
[after PR 6325 recovery] 1 message
>>> 12 of 13 destroyed; 3 fully-valid completed turns lost. No event, no log, no warning.
```

This composes badly with H1: a cancel poisons the transcript, the defect scrolls into mid-history, and the next genuine interruption annihilates everything after it.

---

## M3 — The repair erases the only evidence that a side effect was attempted

The test in this PR asserts the next model input is `[user("This will be interrupted"), user("Continue")]`. The tool call is gone from the model's view. If the interrupted call was a `bash rm -rf …` that had already begun executing, the model has no idea and may reissue it.

The docs change says recovery "does not replay or reconstruct the interrupted side effect" — but *deleting* the record is strictly worse than substituting a synthetic `tool` result reading `"interrupted; outcome unknown"`. That variant is well-formed for the provider **and** keeps the model informed.

---

## M4 — Event log and transcript now disagree

The test asserts `[[3, "tool_started"], [4, "turn_failed"]]`. `tool_started` survives with no `tool_completed`, while the transcript no longer contains the call. Any UI replaying the durable event stream renders a permanently in-flight tool. Worth emitting a terminal `tool_completed` (interrupted) for each stranded call as part of the repair.

---

## Lower severity

- **Compaction boundary loss.** An assistant message can carry `toolCalls` and a `compaction` item in the same `providerContinuation` (`responses-protocol.ts:415`). Truncating that message drops the pruning boundary `toResponsesInput` depends on (`:213`), so the next request replays the full pre-compaction history — cost spike, or a hard context-length 400. Worth confirming whether the provider ever emits compaction alongside function calls.
- **`slice(0, exchangeStart)`** at `:346` / `:354` takes `number | undefined`. The in-batch invariant holds today, but if it ever breaks, `slice(0, undefined)` returns the *entire* transcript — the repair silently degrades to a no-op instead of failing loudly. Worth an explicit invariant.
- **Multi-writer risk.** `initializeFromStore` now rewrites transcripts for every session holding a non-terminal turn. A second process starting while the first is mid-turn would clobber a live session's history. The single-writer assumption predates this PR, but this change escalates the consequence from "turn status" to "history deletion."

---

## Suggested direction

1. Move the repair to `#runTurn`'s transcript assembly (`:1129`) rather than only `initializeFromStore` — it's the one choke point every entry path crosses, so restart, cancel, and close are all covered at once.
2. Excise the stranded batch instead of truncating to it.
3. Prefer synthesized interrupted-tool results over deletion (addresses M3 and M4 together).
4. Log or emit the count of removed messages — right now the repair is completely silent.

---

<details>
<summary><b>Repro 1 — cancel mid-tool strands a call that recovery never repairs</b> (fails on this branch)</summary>

```ts
import { assertEquals } from "@std/assert";
import { toFileUrl } from "@std/path";
import {
  HarnessInteractiveChatService,
  type HarnessInteractivePromptLoopFactory,
} from "../src/interactive-chat-service.ts";
import type {
  HarnessPromptLoopResult,
  RunHarnessTranscriptOptions,
} from "../src/prompt-loop.ts";
import { openSqliteHarnessChatSessionStore } from "../src/sqlite-session-store.ts";

const nextIsoNow = () => {
  let counter = 0;
  return () => {
    counter += 1;
    return `2026-05-27T00:00:${String(counter).padStart(2, "0")}.000Z`;
  };
};

const makeResult = (
  options: RunHarnessTranscriptOptions,
  finalAssistantText: string,
): HarnessPromptLoopResult => ({
  model: options.model ?? "gpt-test",
  finalAssistantText,
  transcript: [
    ...options.transcript,
    { role: "assistant", content: finalAssistantText },
  ],
  modelTurns: 1,
  runState: {} as HarnessPromptLoopResult["runState"],
});

const shape = (transcript: readonly unknown[]) =>
  transcript.map((m) => {
    const msg = m as Record<string, unknown>;
    if (msg.role === "tool") return `tool(${msg.toolCallId})`;
    if (msg.role === "assistant") {
      const calls = (msg.toolCalls ?? []) as { id: string }[];
      return `assistant[${calls.map((c) => c.id).join(",") || "-"}]`;
    }
    return `${msg.role}(${String(msg.content).slice(0, 24)})`;
  });

Deno.test("REDTEAM: cancel mid-tool strands a tool call that recovery never repairs", async () => {
  const path = await Deno.makeTempFile({ suffix: ".sqlite" });
  const store = await openSqliteHarnessChatSessionStore({ url: toFileUrl(path) });

  try {
    let resolveToolCallPersisted: (() => void) | undefined;
    const toolCallPersisted = new Promise<void>((r) => {
      resolveToolCallPersisted = r;
    });

    // Prompt loop: model asks for a tool, the tool runs long, then the turn is
    // aborted. This is an ordinary user-initiated cancel, not a process death.
    const createPromptLoop: HarnessInteractivePromptLoopFactory = () => ({
      runTranscript: async (options) => {
        const assistant = {
          role: "assistant" as const,
          content: "",
          toolCalls: [{
            id: "call-canceled",
            type: "function" as const,
            function: {
              name: "bash",
              arguments: '{"command":"sleep 600"}',
            },
          }],
        };
        await options.onTranscriptEvent?.({
          message: assistant,
          transcript: [...options.transcript, assistant],
        });
        resolveToolCallPersisted?.();
        return await new Promise<HarnessPromptLoopResult>((_, reject) => {
          options.signal?.addEventListener("abort", () => {
            reject(options.signal?.reason ?? new Error("aborted"));
          });
        });
      },
    });

    const service = new HarnessInteractiveChatService({
      createPromptLoop,
      now: nextIsoNow(),
      sessionStore: store,
    });
    await service.startSession("req-1", {
      sessionId: "session-1",
      workspace: { hostPath: "/workspace" },
    });
    await service.startTurn("req-2", {
      sessionId: "session-1",
      turnId: "turn-1",
      input: { text: "Delete the temp files" },
    });
    await toolCallPersisted;

    await service.cancelTurn("req-3", "session-1", "turn-1", "user_requested");
    await service.waitForTurn("session-1", "turn-1");

    console.log(
      "\n[1] turn status after cancel:",
      service.listTurns({ sessionId: "session-1" }).turns.map((t) =>
        t.turn.status
      ),
    );
    console.log(
      "[2] durable transcript after cancel:",
      shape(store.getSession("session-1")?.transcript ?? []),
    );
    console.log(
      "[3] session reusable:",
      service.status("session-1").sessions[0].reusable,
    );

    // Simulate a full process restart -> this is where PR 6325's repair runs.
    const restoredInputs: RunHarnessTranscriptOptions["transcript"][] = [];
    const restored = new HarnessInteractiveChatService({
      createPromptLoop: () => ({
        runTranscript: (options) => {
          restoredInputs.push([...options.transcript]);
          return Promise.resolve(makeResult(options, "Recovered."));
        },
      }),
      now: () => "2026-05-27T00:01:00.000Z",
      sessionStore: store,
    });
    await restored.initializeFromStore();

    console.log(
      "[4] durable transcript AFTER PR 6325 recovery:",
      shape(store.getSession("session-1")?.transcript ?? []),
    );

    await restored.startTurn("req-4", {
      sessionId: "session-1",
      turnId: "turn-2",
      input: { text: "Continue" },
    });
    await restored.waitForTurn("session-1", "turn-2");

    console.log(
      "[5] transcript sent to the model on the NEXT turn:",
      shape(restoredInputs[0] ?? []),
    );

    const next = restoredInputs[0] ?? [];
    const calls = new Set(
      next.flatMap((m) =>
        m.role === "assistant" ? (m.toolCalls ?? []).map((c) => c.id) : []
      ),
    );
    const results = new Set(
      next.flatMap((m) => m.role === "tool" ? [m.toolCallId] : []),
    );
    const unmatched = [...calls].filter((id) => !results.has(id));
    console.log("[6] UNMATCHED tool calls in next model input:", unmatched);

    assertEquals(
      unmatched,
      [],
      `next model request replays unmatched tool call(s) -> provider HTTP 400`,
    );
  } finally {
    store.database.close();
    await Deno.remove(path).catch(() => {});
  }
});
```

</details>

<details>
<summary><b>Repro 2 — repair truncates at the first defect, destroying all later history</b> (passes, printing the loss)</summary>

```ts
import { toFileUrl } from "@std/path";
import {
  createHarnessChatSessionStatus,
  DEFAULT_HARNESS_CHAT_POLICY,
} from "../src/contracts/interactive-chat.ts";
import type { HarnessChatTurnRecord } from "../src/contracts/interactive-chat.ts";
import type { HarnessTranscriptMessage } from "../src/contracts/transcript.ts";
import {
  HarnessInteractiveChatService,
  type HarnessInteractivePromptLoopFactory,
} from "../src/interactive-chat-service.ts";
import type { HarnessPromptLoopResult } from "../src/prompt-loop.ts";
import { openSqliteHarnessChatSessionStore } from "../src/sqlite-session-store.ts";

const call = (id: string) => ({
  id,
  type: "function" as const,
  function: { name: "bash", arguments: "{}" },
});
const shape = (t: readonly HarnessTranscriptMessage[]) =>
  t.map((m) =>
    m.role === "tool"
      ? `tool(${m.toolCallId})`
      : m.role === "assistant"
      ? `assistant[${(m.toolCalls ?? []).map((c: { id: string }) => c.id).join(",") || "-"}]`
      : `${m.role}(${m.content.slice(0, 18)})`
  );

Deno.test("REDTEAM: repair truncates at the FIRST defect, destroying all later history", async () => {
  const path = await Deno.makeTempFile({ suffix: ".sqlite" });
  const store = await openSqliteHarnessChatSessionStore({ url: toFileUrl(path) });

  try {
    // A session whose transcript carries one stranded call early on (e.g. left
    // by a cancel), followed by many complete, valid turns.
    const transcript: HarnessTranscriptMessage[] = [
      { role: "user", content: "turn 1 question" },
      { role: "assistant", content: "", toolCalls: [call("stranded-early")] },
      { role: "user", content: "turn 2 question" },
      { role: "assistant", content: "", toolCalls: [call("ok-2")] },
      { role: "tool", toolCallId: "ok-2", toolName: "bash", content: "{}" },
      { role: "assistant", content: "turn 2 answer" },
      { role: "user", content: "turn 3 question" },
      { role: "assistant", content: "", toolCalls: [call("ok-3")] },
      { role: "tool", toolCallId: "ok-3", toolName: "bash", content: "{}" },
      { role: "assistant", content: "turn 3 answer" },
      { role: "user", content: "turn 4 question" },
      { role: "assistant", content: "turn 4 answer" },
      { role: "user", content: "turn 5 question" },
    ];

    const status = createHarnessChatSessionStatus({
      sessionId: "session-1",
      workspace: { hostPath: "/workspace" },
      createdAt: "2026-05-27T00:00:00.000Z",
    });
    // Turn 5 is the one that was in flight when the process died.
    const running: HarnessChatTurnRecord = {
      sessionId: "session-1",
      turn: {
        turnId: "turn-5",
        status: "running",
        startedAt: "2026-05-27T00:00:00.000Z",
        updatedAt: "2026-05-27T00:00:00.000Z",
      },
      input: { text: "turn 5 question" },
      policy: DEFAULT_HARNESS_CHAT_POLICY,
    };
    store.saveSession({
      session: { ...status, status: "turn_running", activeTurnId: "turn-5" },
      transcript,
    });
    store.saveTurn(running);

    console.log(`\n[before] ${transcript.length} messages:`);
    console.log("  ", shape(transcript).join(" "));

    const service = new HarnessInteractiveChatService({
      createPromptLoop: (() => ({
        runTranscript: () => Promise.resolve({} as HarnessPromptLoopResult),
      })) as HarnessInteractivePromptLoopFactory,
      now: () => "2026-05-27T00:01:00.000Z",
      sessionStore: store,
    });
    await service.initializeFromStore();

    const after = store.getSession("session-1")?.transcript ?? [];
    console.log(`\n[after PR 6325 recovery] ${after.length} messages:`);
    console.log("  ", shape(after).join(" ") || "(empty)");
    console.log(
      `\n>>> ${transcript.length - after.length} of ${transcript.length} messages destroyed; ` +
        `3 fully-valid completed turns lost. No event, no log, no warning.`,
    );
  } finally {
    store.database.close();
    await Deno.remove(path).catch(() => {});
  }
});
```

</details>

**Comment by tonyespinoza1** [comment]
Red-team follow-up resolved at `81ca9d3e2`:

- Deleted `truncateIncompleteToolExchange`; one non-destructive normalizer now runs at durable restore and immediately before the next turn.
- Missing results become explicit `outcome: "unknown"` tool results. The payload has no `ok: false` (or success claim) and says that execution and side effects are unknown.
- Preserves the assistant call, partial completed results, later history, and provider compaction continuation.
- Orphan/duplicate results and duplicate call IDs fail closed locally before model traffic; the durable transcript remains unchanged.
- Added persistence-boundary idempotence proof using SQLite's total-change count, plus cancel-mid-tool, partial parallel batch, malformed legacy, complete no-op, and restart coverage.

Event decision: this does not synthesize `tool_completed`. That event's current status vocabulary is `completed | failed | denied`, so every available value would overclaim an unknown outcome. The real turn still ends through its existing `turn_failed` or `turn_canceled` event; an explicit interrupted/unknown tool-event status is adjacent contract/UI work, not hidden in this repair.

Local exact-head validation: cf-harness 767 tests / 444 steps, all 43 type-check groups, all 566 docs blocks, changed-file lint/format, and diff check. CI and exact-head Cubic review are in progress; I will not merge.

Codex task: 🛠️ Fixing Labs PR 6325 — `01a03b44-52b8-7c63-8ff3-da8157912965`

**Comment by tonyespinoza1** [comment]
Cubic follow-up on the exact-head event-status finding:

The synthetic unknown-outcome result does not reach toolMessageStatus and is not emitted as tool_completed. CfHarnessPromptLoop replays its initial transcript through onTranscriptEvent, but HarnessInteractiveChatService initializes observedTranscriptLength to that full input length and ignores every replay callback at or below the boundary. Only messages newly appended during the current turn are emitted as current-turn activity.

Head 78b034146 makes that contract explicit and extends the SQLite recovery regression to perform the initial replay, append a real new assistant message, and assert the event stream is exactly turn_started, assistant_delta, assistant_completed, turn_completed — with no tool_started/tool_completed for historical recovery evidence. The focused SQLite suite passes 16/16.

Adding an unknown public event status would therefore widen the client contract for an event this path intentionally does not emit.

File: packages/cf-harness/src/interactive-chat-service.ts [hunk 1]
```diff
@@ -330,12 +331,172 @@ const interruptedTurnError = (
...
+    }
+  }
+
+  const insertions = new Map<number, HarnessToolCall[]>();
+  for (const { toolCall, transcriptIndex } of pendingToolCalls.values()) {
+    // Keep a batch's real contiguous results in their recorded order, then
+    // close only its missing calls before later conversation or compaction can
+    // move the provider projection boundary past the declaring call.
+    let insertionIndex = transcriptIndex;
+    while (transcript[insertionIndex + 1]?.role === "tool") {
```
[the remark this answers is not in the document]
Comment (tonyespinoza1) [reply packages/cf-harness/src/interactive-chat-service.ts:469]: Reasoned reply: I do not think this finding is valid under the transcript contract this PR is preserving. `inspectHarnessTranscriptPairing` explicitly pairs by ID across the whole transcript and its contract test accepts a result after an intervening user message. The prior Cubic finding on this head family correctly required the normalizer to stop imposing adjacency for that reason.

Moving a missing sibling result to the last non-adjacent real-result index is unsafe when later history carries a Responses compaction boundary: the projection can prune the declaring call while retaining the synthesized output, recreating the provider-invalid orphan. The red-first `sqlite session restore preserves partial batches later history and compaction` regression exercises that case and asserts the projected function-call and function-call-output ID sets remain equal.

The current placement keeps a synthetic unknown result on the declaring call side of any later compaction boundary; a genuine late result remains where it was recorded and still pairs by ID. The normalized full transcript and the projected provider input are both paired. I am therefore leaving the implementation unchanged and requesting re-review with this contract/projection evidence.

— Codex task “🔗 PR 6325 review fixes posted” (Thread ID `01a03b44-52b8-7c63-8ff3-da8157912965`)

File: packages/cf-harness/src/interactive-chat-service.ts [hunk 2]
```diff
@@ -310,6 +311,145 @@ const interruptedTurnError = (
...
+
+const unknownToolOutcome = (
+  toolCall: HarnessToolCall,
+): HarnessToolTranscriptMessage => ({
+  role: "tool",
+  toolCallId: toolCall.id,
+  toolName: toolCall.function.name,
+  content: JSON.stringify({
+    type: "cf-harness.tool-outcome-unknown",
+    outcome: "unknown",
```
[the remark this answers is not in the document]
Comment (tonyespinoza1) [reply packages/cf-harness/src/interactive-chat-service.ts:393]: Addressed in 78b0341463f955afbd7959f4716889c37e8a81f8. Recovered durable history is deliberately suppressed during the prompt-loop initial replay via the observed-transcript-length guard, so a synthesized unknown-outcome result is never re-emitted as a newly completed tool call. The regression asserts the post-restore event sequence contains no tool_completed event. Expanding the public event-status enum is therefore unnecessary. — Codex task 🛠️ Auditing Cubic threads on PR 6325, Thread ID 01a03b44-52b8-7c63-8ff3-da8157912965

File: packages/cf-harness/src/interactive-chat-service.ts [hunk 3]
```diff
@@ -800,66 +956,57 @@ export class HarnessInteractiveChatService {
...
+      // own status event. Adoption still happens only after the write commits.
+      try {
         await this.#emit(record.status.sessionId, undefined, {
           kind: "status_changed",
           session: { ...record.status, updatedAt },
-        }, { transcript: recovered.transcript });
-        continue;
+        }, { transcript: normalizedTranscript });
+      } catch (error) {
+        record.recoveryError = normalizationPersistenceError(
```
[the remark this answers is not in the document]
Comment (tonyespinoza1) [reply packages/cf-harness/src/interactive-chat-service.ts]: Addressed in eccb1c36c4807cdc236b7d9766750eab1e0863ae. The catch now sets recoveryError only when the normalized transcript was not adopted, which distinguishes store failure from a listener exception after commit. The regression a listener failure after normalization commit does not poison the session proves the durable transcript remains reusable. — Codex task 🛠️ Auditing Cubic threads on PR 6325, Thread ID 01a03b44-52b8-7c63-8ff3-da8157912965

File: packages/cf-harness/src/interactive-chat-service.ts [hunk 4]
```diff
@@ -330,12 +331,177 @@ const interruptedTurnError = (
...
+      }
+      normalized.push(unknownToolOutcome(toolCall));
+      synthesizedToolCallIds.push(toolCall.id);
+    }
+    pendingToolCalls = [];
+    pendingToolCallIds = new Set();
+  };
+
+  for (const [index, message] of transcript.entries()) {
+    if (pendingToolCalls.length > 0) {
```
[the remark this answers is not in the document]
Comment (tonyespinoza1) [reply packages/cf-harness/src/interactive-chat-service.ts]: Addressed in eccb1c36c4807cdc236b7d9766750eab1e0863ae. The normalizer now pairs calls and results by ID across the entire transcript rather than adjacency. The sqlite session restore pairs late tool results across interleaved history regression preserves the valid transcript unchanged. — Codex task 🛠️ Auditing Cubic threads on PR 6325, Thread ID 01a03b44-52b8-7c63-8ff3-da8157912965

File: packages/cf-harness/docs/CURRENT_STATE.md [hunk 5]
```diff
@@ -256,9 +256,14 @@ mode.
   explicitly preloaded by the caller; child skills are profile-controlled.
 - Resume is transcript-oriented and does not recover an arbitrary partially
   executed tool or orchestration state machine. An interrupted turn is never
-  replayed automatically: its tools may already have taken effect, so the
-  session resumes from the checkpoint before it and the operator decides what to
-  send next.
+  replayed automatically: new turns retain their last provider-safe durable
+  checkpoint, while restore and the next-turn boundary normalize legacy
```
[the remark this answers is not in the document]
Comment (tonyespinoza1) [reply packages/cf-harness/docs/CURRENT_STATE.md:260]: Addressed in eccb1c36c4807cdc236b7d9766750eab1e0863ae. The Supported surfaces paragraph now documents non-destructive unknown-outcome synthesis and local refusal for orphan results or duplicate call IDs; the obsolete resumable-prefix truncation language is gone. — Codex task 🛠️ Auditing Cubic threads on PR 6325, Thread ID 01a03b44-52b8-7c63-8ff3-da8157912965

File: packages/cf-harness/src/interactive-chat-service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tonyespinoza1) [reply packages/cf-harness/src/interactive-chat-service.ts:469]: Disposition confirmed after the requested regression. 7a68150db319b8774db8100b5a3c3c1d23491201 adds sqlite restore closes a missing sibling without moving a late real result, proving both the full transcript and provider projection remain paired while the synthetic result stays on the declaration side of later compaction. Moving it downstream as suggested would create an orphan after compaction. Exact-head Cubic check 98219369879 subsequently reported 0 issues across 3 files. — Codex task 🛠️ Auditing Cubic threads on PR 6325, Thread ID 01a03b44-52b8-7c63-8ff3-da8157912965
